### ISSUE TITLE
GetInputAsJson doesn't return a JToken representative of the payload

### ISSUE THREAD

**Issue by brandonh-msft** [Azure/azure-functions-durable-extension#247]
# Problem statement
GetInputAsJson on an Activity Function doesn't return a JToken representative of the payload sent to it from an Orchestrator

## Tooling Versions
* VS 2017: 15.6.6
* Functions Runtime: v1.0.10
* Functions Extension for VS: 15.0.40405
* Functions nupkg: v1.0.13
* DurableTask Extension nupkg: 1.2.2-beta3

## Observed behavior
In my [serverless-file-validation](https://github.com/brandonh-msft/serverless-file-validation) sample, I have some code that does this:
```csharp
await context.CallActivityAsync(@"ValidateFileSet", new { prefix = $@"{newCustomerFile.ContainerName}/inbound/{newCustomerFile.BatchPrefix}", fileTypes = expectedFiles });
```
and then in the receiving activity I have this:
```csharp
var payload = context.GetInputAsJson();
```
What comes through is an object with a root *array* that is the **string value** of the JSON you sent. In other words you have to do this to get out what I expected:
```csharp
payload = Newtonsoft.Json.Linq.JToken.Parse(payload.Root[0].ToString());
```
Which seems not only unintuitive, but very cumbersome.
 
I also tried creating a manual JSON payload (instead of the Anonymous Type) in the `CallActivityAsync` line like so:
```csharp
var jsonObjectForValidator =
$@"{{
  ""prefix"" : ""{newCustomerFile.ContainerName}/inbound/{newCustomerFile.BatchPrefix}"",
  ""fileTypes"" : [{string.Join(", ", expectedFiles.Select(e => $@"""{e}"""))}]
}}";
// call next step in functions with the prefix so it knows what to go grab
await context.CallActivityAsync(@"ValidateFileSet", jsonObjectForValidator);
```
 
but got the same result.

## Expected Behavior
If you look at the signature of `GetInputAsJson` you see you get back a `JToken`. I expect, in this instance a `JToken` where I could say `["prefix"].ToString()` or `Value<string>("prefix")` and get out the value of the `prefix` property I put in.

## Workarounds
Use this code instead:
```csharp
var payload = context.GetInputAsJson();
payload = Newtonsoft.Json.Linq.JToken.Parse(payload.Root[0].ToString());
```

## Other commentary
The curious thing is that it [looks like](https://github.com/Azure/azure-functions-durable-extension/blob/master/src/WebJobs.Extensions.DurableTask/DurableActivityContext.cs#L57) this shouldn't be happening at all:
```csharp
public JToken GetInputAsJson()
{
    if (this.serializedInput != null && this.parsedJsonInput == null)
    {
        JArray array = JArray.Parse(this.serializedInput);
        if (array?.Count != 1)
        {
            throw new ArgumentException("The serialized input is expected to be a JSON array with one element.");
        }

        this.parsedJsonInput = array[0];
    }

    return this.parsedJsonInput;
}
```

So I'm not sure if this just a bug in the currently-available nupkg, or if something more nefarious is going on here.

**Comment by brandonh-msft** [comment]
After debugging as much of the SDK as I can, it looks like this is caused by the queuing of the payload in `StartNewAsync` or `RaiseEvent`, and not the retrieving of it. This queuing is not taking in to account a JSON-formatted string value; it treats that as simply a string, and ends up wrapping it in a JSON body when you `JArray.Parse` it.

Unfortunately, the code that ultimate does this appears to be in `DurableFunctions.Core` which we don't have the source for. I'll leave this issue open and propose - for consideration - a couple of different possible ways to solve this on the retrieval end in case we want to go that route, but ultimately I think this should be corrected where we queue the payload.

**Comment by cgillum** [comment]
I'm aware of the behavior in DurableTask.Core which puts the parameter into an array (not sure why they did this - I suspect it was a point-in-time thing), but I thought the code in our layer properly compensated for this - i.e. the code you pointed out.   Do you have a simple activity/orchestration pair that you can share which reproduces the issue? I'd like to take a look at it under a debugger.

**Comment by brandonh-msft** [comment]
Nothing "simple" (can repro at [this commit point in my repo](https://github.com/brandonh-msft/serverless-file-validation/tree/bda30bf471ed7b6d626b28e1ac9489b5a907a4b3)) but I'll create and attach.

**Comment by brandonh-msft** [comment]
Reproduces all scenarios:
```csharp
public static class Function1
{
    [FunctionName("starter")]
    public static async Task<HttpResponseMessage> Starer([HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = null)]HttpRequestMessage req, [OrchestrationClient]DurableOrchestrationClient starter, TraceWriter log)
    {
        log.Info("C# HTTP trigger function processed a request.");


        var instance = await starter.StartNewAsync("runner", null);

        return starter.CreateCheckStatusResponse(req, instance);
    }

    [FunctionName("runner")]
    public static async Task Runner([OrchestrationTrigger]DurableOrchestrationContext context, TraceWriter log)
    {
        await context.CallActivityAsync(@"activity", @"{ ""myvalue"" : ""value1"" }");

        await context.CallActivityAsync(@"objActivity", new { myvalue = @"value1" });

        await context.CallActivityAsync(@"workingStringActivity", @"{ ""myvalue"" : ""value1"", ""myvalue2"" : ""value2"" }");
    }

    [FunctionName(@"activity")]
    public static bool Activity([ActivityTrigger]DurableActivityContext context, TraceWriter log)
    {
        try
        {
            var payload = context.GetInputAsJson();
            var myvalue = payload["myvalue"].ToString(); // Expecting 'value1' here, will blow up
            log.Warning($@"payload 'myvalue': {myvalue} ");

            return true;
        }
        catch (Exception e)
        {
            log.Error(e.Message, e);
            return false;
        }
    }

    [FunctionName(@"objActivity")]
    public static bool ObjectActivity([ActivityTrigger]DurableActivityContext context, TraceWriter log)
    {
        var payload = context.GetInputAsJson();
        var myvalue = payload["myvalue"].ToString(); // Expecting 'value1' here, will work
        log.Warning($@"payload 'myvalue': {myvalue} ");

        return true;
    }

    [FunctionName(@"workingStringActivity")]
    public static bool WorkingStringActivity([ActivityTrigger]DurableActivityContext context, TraceWriter log)
    {
        var payload = context.GetInputAsJson();
        var myvalue = JToken.Parse(payload.Root[0].ToString())["myvalue"]; // Expecting 'value1' here, will work
        log.Warning($@"payload 'myvalue': {myvalue.Value<string>()} ");

        return true;
    }
}
```

**Comment by brandonh-msft** [comment]
Right; I wonder if this could be solved by either 1) more clearly defining what we're doing with the `object` overload (serializing it to JSON) or 2) requiring *all* payload to be JSON (serializable and/or parseable)

I guess the question we'd have to ask our users is: do you regularly pass primitives as parameters to `CallActivityAsync` and fetch them out as their type, or are you regularly JSON-encoding everything and using it that way?

If it's the latter, I'd rather our signature enforce JSON-compatibility for things. Might be something to think about?

**Comment by cgillum** [comment]
We do call out [in the API documentation](https://azure.github.io/azure-functions-durable-extension/api/Microsoft.Azure.WebJobs.DurableOrchestrationContext.html#Microsoft_Azure_WebJobs_DurableOrchestrationContext_CallActivityAsync_System_String_System_Object_) that the parameters need to be serializeable. Did you have something more specific in mind?

I don't think there is a way the signature can enforce JSON serializeability at compile time. What we do now is simply throw an exception at runtime if the payload can't be serialized. In my experience with the .NET framework, it's not uncommon to do this. Usually when I see overloads that take primitives, it's for performance reasons - i.e. use this overload of this low-level API to avoid CLR boxing operations. That's less of a concern in our case since the JSON serializer is likely going to need to box the value anyways (and the overhead of serialization which make those low-level optimizations somewhat meaningless).

The confusion about how to send a collection of parameters is valid and I've heard this from others. My thinking was to add documentation to explicitly cover this: https://github.com/Azure/azure-functions-durable-extension/issues/152

**Comment by cgillum** [comment]
The two main problems I have with this proposal are:

1. This greatly lowers usability of CallActivityAsync and related APIs
2. This would be a massive breaking change.

I honestly don't think it is worth it to simply make `GetInputAsJson` more intuitive. If anything, I would rather get rid of `GetInputAsJson` and possibly `GetRawInput` to avoid confusion.

Ultimately I don't want developers to have to think in terms of JSON. They should only need to think in terms of their app specific APIs and logic. The object abstraction also gives us the freedom to change serialization formats in the future if we prefer (for example, more efficient binary formats, or JSON formats that don't depend on the 3rd party `Newtonsoft.Json` package).

**Comment by brandonh-msft** [comment]
I'd also be for getting rid of those two methods :) & totally understand the point about flexibility to change serialization formats & dependencies.

Good convo - thanks! 👍
